lls/cache.h

```diff
@@ -0,0 +1,128 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _GATEKEEPER_LLS_CACHE_H_
+#define _GATEKEEPER_LLS_CACHE_H_
+
+#include "gatekeeper_lls.h"
+#include "gatekeeper_net.h"
+
+/*
+ * Maximum key length for an LLS cache entry, used to determine
+ * how much memory should be allocated for cache requests.
+ * Therefore, it should be set to the maximum key length
+ * needed, but no more to avoid wasting mempool allocation.
+ */
+#define LLS_CACHE_MAX_KEY_LEN	4
+
+enum lls_cache_op {
+	/* Add an entry to the cache. */
+	LLS_CACHE_ADD,
+	/* Increment the refcnt of a cache entry. */
+	LLS_CACHE_INC,
+	/* Decrement the refcnt of a cache entry. */
+	LLS_CACHE_DEC,
+};
+
+struct lls_cache_req {
+	/* Entry to be stored. */
+	struct lls_cache_entry *entry;
+
+	/* Operation of this request. */
+	enum lls_cache_op      op;
+
+	/* The cache that this request will change. */
+	struct lls_cache       *cache;
+
+	/*
+	 * IP address for a new entry (or key for the hash).
+	 *
+	 * NOTE
+	 *	This is a variable-length field.
+	 */
+	const void             *ip_key[0];
+};
+
+struct lls_cache_entry {
+	/* Ethernet address of this entry. */
+	struct ether_addr ha;
+
+	/*
+	 * Time that entry was added. When the entry is updated,
+	 * the timestamp is also updated.
+	 */
+	time_t            ts;
+
+	/* Port on which this entry exists. */
+	uint8_t           port_id;
+
+	/* Whether this entry has been marked as stale. */
+	int               stale;
+
+	/*
+	 * Whether this entry has a valid resolution. This is false
+	 * under two conditions:
+	 *   (1) If a block has expressed interest in an entry but
+	 *       it has not yet been resolved.
+	 *   (2) If a resolution for this entry becomes stale and
+	 *       requests are not answered but there is still
+	 *       interest in this entry.
+	 */
+	int               valid;
```

Comment:
Good simplification.

I think that the scanning code should take advantage of unicast transmissions. But instead of having a state field for the entry, we could have a state field to scanning code. Example, when the entry becomes stale (or before it becomes stale), a unicast packet is sent. If the scanning code still finds a stale entry after a unicast transmission has been sent previously, then it issues a broadcast packet.

The motivation for this suggestion is that, in production, many of the entry will be validated by the unicast packets, and avoiding broadcast reduces load on the physical network.